src/udp.rs

```diff
@@ -39,103 +114,293 @@ impl UdpSocket {
+
+            // set the new destination after we've successfully set them on chip
+            self.destination = Some(remote);
+        }
+
+        Ok(())
+    }
+
+    fn socket_send_all<SpiBus: Bus>(
+        &self,
+        bus: &mut SpiBus,
+        send_buffer: &[u8],
+    ) -> NbResult<(), UdpSocketError<SpiBus::Error>> {
+        match Status::try_from(self.socket.get_status(bus)?) {
+            Ok(Status::Udp) => {}
+            Ok(status) => return Err(NbError::Other(UdpSocketError::SocketNotOpen)),
+            Err(err) => return Err(NbError::Other(UdpSocketError::UnrecognisedStatus)),
+        }
+
+        if self.destination.is_none() {
+            return Err(NbError::Other(UdpSocketError::DestinationNotSet));
+        }
+
+        let mut free_size = self.socket.get_tx_free_size(bus)?;
+
+        // Ensure write is currently possible.
+        // This should never be `0`
+        if free_size == 0 {
+            // If this happens, then `Send` was not called in previous operations.
+            // FIXME: add a way to either:
+            // - flush the buffer by clearing it up
+            // - `Send` the data to its the destination - this might not be possible if the destination has changed.
+            return Err(NbError::Other(UdpSocketError::BufferFull));
+        }
+
+        // on the first send, we rely on the free TX size and the length of the send_buffer
+        let mut total_sent = 0;
+
+        // if we've sent all bytes, exit the loop
+        while total_sent != send_buffer.len() {
+            let remaining_bytes = &send_buffer[total_sent..];
+
+            // take the smaller value out of the free buffer length and
+            // the length of the remaining bytes to be send
+            let write_len = (free_size as usize).min(remaining_bytes.len());
+            let send_batch = &remaining_bytes[..write_len];
+
+            // but on consequent send calls, we leave that to send
+            let sent = self.socket_send(bus, send_batch)?;
+
+            total_sent += sent;
+
+            // update the `free_size` of the TX buffer after we've sent some bytes
+            free_size = self.socket.get_tx_free_size(bus)?;
+        }
+
         Ok(())
     }
 
+    /// Send buffer should be less than [`u16::MAX`].
+
+    /// Will block the device until all data is sent.
+    ///
+    /// # Returns
+    ///
+    /// The amount of bytes that were sent. The caller should make sure to
+    /// check and send the rest of the `send_buffer` data.
     fn socket_send<SpiBus: Bus>(
         &self,
         bus: &mut SpiBus,
         send_buffer: &[u8],
+    ) -> NbResult<usize, UdpSocketError<SpiBus::Error>> {
+        match Status::try_from(self.socket.get_status(bus)?) {
+            Ok(Status::Udp) => {}
+            Ok(status) => return Err(NbError::Other(UdpSocketError::SocketNotOpen)),
+            Err(err) => return Err(NbError::Other(UdpSocketError::UnrecognisedStatus)),
+        }
```

Comment:
This is copy/pasted at least 3 different times. I think the best way would be to update the `Status` API so we can write this like:
```suggestion
        self.socket.get_status(bus)?.as_udp_result()?;
```

Or something similar. The ergonomics here aren't great as-written